bouncer/shared/utils/chainflip_io.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  ChainflipExtrinsicSubmitter,
+  createStateChainKeypair,
+  extractExtrinsicResult,
+  lpMutex,
+} from 'shared/utils';
+import { z } from 'zod';
+// eslint-disable-next-line no-restricted-imports
+import type { KeyringPair } from '@polkadot/keyring/types';
+import { DisposableApiPromise, getChainflipApi } from './substrate';
+import { ChooseSingleEvent, EventDescription, EventName, findOneEventOfMany } from './indexer';
+import { Logger } from './logger';
+
+export class ChainflipIO<Requirements> {
+  /// The last block height at which either an input or an output operation happened,
+  /// (that is either an extrinsic was submitted or an event was found)
+  private lastIoBlockHeight: number;
+
+  /// Methods can contain additional requirements that they have, that is additional
+  /// data that should be available when calling them. For example, submitting an
+  /// extrinsic requires a statechain account.
+  readonly requirements: Requirements;
+
+  /// This class also exposes logger functionality
+  readonly logger: Logger;
+
+  /// Creates a new instance, the `lastIoBlockHeight` has to be specified. If you want
+  /// to automatically initialize to the current block height, use `newChainflipIO` instead.
+  constructor(logger: Logger, requirements: Requirements, lastIoBlockHeight: number) {
+    this.lastIoBlockHeight = lastIoBlockHeight;
+    this.requirements = requirements;
+    this.logger = logger;
+  }
+
+  async stepToExtrinsicIncluded<Data extends Requirements & { account: FullAccount<AccountType> }>(
+    this: ChainflipIO<Data>,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    extrinsic: (api: DisposableApiPromise) => any,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ): Promise<Result<any, string>> {
+    await using chainflip = await getChainflipApi();
```

Comment:
Did you consider making the api object a member of this rather than getting/disposing of one in each of the methods?